bindings/rust/s2n-tls/src/testing/s2n_tls.rs

```diff
@@ -0,0 +1,325 @@
+    unsafe extern "C" fn recv_cb(context: *mut c_void, data: *mut u8, len: u32) -> s2n_status_code {
+        let context = &mut *(context as *mut Self);
+        let data = core::slice::from_raw_parts_mut(data, len as _);
+        match context.on_read(data) {
+            0 => {
+                // https://github.com/awslabs/s2n/blob/main/docs/USAGE-GUIDE.md#s2n_connection_set_send_cb
+                // s2n-tls wants us to set the global errno to signal blocked
+                errno::set_errno(errno::Errno(libc::EWOULDBLOCK));
+                -1
+            }
+            len => len as _,
+        }
+    }
+
+    /// Called when receiving data
+    fn on_read(&mut self, data: &mut [u8]) -> usize {
+        let max_len = Some(data.len());
+
+        if let Some(chunk) = self.context.receive(max_len) {
+            let len = chunk.len();
+            //dbg!(len);
+            data[..len].copy_from_slice(&chunk);
+            len
+        } else {
+            0
+        }
+    }
+}
+
+pub mod tests {
+    use super::*;
+    use crate::raw::{config::*, error::Error};
+    use crate::testing::Pair;
+
+    const SAMPLES: usize = 100;
+
+    #[allow(unused_variables)]
+    struct HostVerifyData {
+        callback_invoked: bool,
+        allow: bool,
+        host: &'static str,
+    }
+    struct CertKeyPair {
+        cert: &'static [u8],
+        key: &'static [u8],
+    }
+    #[allow(unused_variables, dead_code)]
+    impl CertKeyPair {
+        fn default() -> Self {
+            CertKeyPair {
+                cert: include_str!("../../../../../tests/pems/rsa_4096_sha512_client_cert.pem")
+                    .as_bytes(),
+                key: include_str!("../../../../../tests/pems/rsa_4096_sha512_client_key.pem")
+                    .as_bytes(),
+            }
+        }
+        fn ecdsa() -> Self {
+            CertKeyPair {
+                cert: include_str!("../../../../../tests/pems/ecdsa_p256_pkcs1_cert.pem")
+                    .as_bytes(),
+                key: include_str!("../../../../../tests/pems/ecdsa_p256_pkcs1_key.pem")
+                    .as_bytes(),
+            }
+        }
+        fn cert(&mut self) -> &'static [u8] {
+            self.cert
+        }
+        fn key(&mut self) -> &'static [u8] {
+            self.key
+        }
+    }
+
+    pub fn build_config(cipher_prefs: &str) -> Result<crate::raw::config::Config, Error> {
+        let mut builder = Builder::new();
+        let mut keypair = CertKeyPair::default();
+        // Build a config
+        builder
+            .set_cipher_preference(cipher_prefs)
+            .expect("Unable to set config cipher preferences");
+        builder
+            .load_pem(keypair.cert(), keypair.key())
+            .expect("Unable to load cert/pem");
+        unsafe {
+            let mut ctx = HostVerifyData {
+                allow: true,
+                callback_invoked: false,
+                host: "127.0.0.1",
+            };
```

Comment:
This needs to be boxed before passing it to a callback. The way it is now, you're reading arbitrary memory on the stack. I'm surprised it's working at all 😨